### PR TITLE
Introduce `createForm` to replace `createControlGroup`

### PR DESCRIPTION
**Pull request by cychiuae**

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -274,3 +275,24 @@ export interface PickerProps {
 }
 
 export class Picker extends React.Component<PickerProps> {}
+
+declare class Form extends React.Component<ScrollViewProps> {}
+
+interface FormFieldRenderProps {
+  focusableRef: React.Ref<any>;
+  onSubmitEditing: (e: any) => void;
```
Comment (louischan-oursky) [inline index.d.ts]: I think React Native should have export a proper type, maybe we can use it.

File: src/Form/createForm.tsx [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+    };
+
+    render() {
+      const { children, ...restProps } = this.props;
+
+      return (
+        <Provider value={this.state.context}>
+          <ScrollView
+            ref={this.scrollViewRef}
+            alwaysBounceVertical={false}
```
Comment (louischan-oursky) [inline src/Form/createForm.tsx]: Any reason for this? It can be overridden by restProps.

File: src/Form/createForm.tsx [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+            (_x, y) => {
+              const { scrollViewContentHeight, scrollViewHeight } = this.state;
+              if (scrollViewContentHeight <= scrollViewHeight) {
+                return;
+              }
+              const maxContentOffsetY =
+                this.state.scrollViewContentHeight -
+                this.state.scrollViewHeight;
+              const desiredContentOffsetY = Math.min(
+                y < 100 ? 0 : (y * 2) / 3,
```
Comment (louischan-oursky) [inline src/Form/createForm.tsx]: Suppose we want to place the text input in the center of the scroll view, the content offset y we want should be `y + (H - h) / 2` where `H` is the scroll view height and `h` is the height of the text input.

File: src/Form/createForm.tsx [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+    focusNext: (index: number) => void;
+    setFieldInstance: (
+      index: number,
+      focusable: FocusableContainer | null
+    ) => void;
+  };
+
+  scrollViewHeight: number;
+  scrollViewContentHeight: number;
+  containerMinHeight: number;
```
Comment (louischan-oursky) [inline src/Form/createForm.tsx]: Leave a comment to explain what purpose `containerMinHeight`serves.

File: src/createControlGroup.tsx [hunk 5]
```diff
@@ -46,6 +46,10 @@ const defaultContextValue: ContextValue = {
 };
 
 export default function createControlGroup() {
+  if (__DEV__) {
+    console.warn("Maybe try to use createForm?");
```
Comment (louischan-oursky) [inline src/createControlGroup.tsx]: Use this message

> createForm integrates ScrollView and auto scroll to TextInput. Maybe you want that instead.

File: index.d.ts (hunk 1 above)
Comment (cychiuae) [reply index.d.ts]: Sure. Let me check check

File: src/Form/createForm.tsx (hunk 2 above)
Comment (cychiuae) [reply src/Form/createForm.tsx]: The purpose of this is I dont want the form can be scrolled if the content size is smaller than the scrollview size. Imagine there is only 1 input field and if the form can be scrolled, I think that is weird UX.

File: src/createControlGroup.tsx (hunk 5 above)
Comment (cychiuae) [reply src/createControlGroup.tsx]: Sure

File: src/Form/createForm.tsx (hunk 4 above)
Comment (cychiuae) [reply src/Form/createForm.tsx]: ok

File: src/Form/createForm.tsx (hunk 3 above)
Comment (cychiuae) [reply src/Form/createForm.tsx]: That will be weird ff the input positioned in centre of scrollview. 
Let's introduce a `scrollToInputThresholds` which does two things. 
1. If `textInput.y` is smaller the the thresholds, we will not scroll the scroll view. 
2. We will scroll to `textInput.y - scrollToInputThresholds`, which will simplify our math and life
Or, additionally we can have delegate for user calculating their desire scroll position

What do you think?

File: src/Form/createForm.tsx (hunk 2 above)
Comment (louischan-oursky) [reply src/Form/createForm.tsx]: It would be great if you can put this rationale in comment.

File: src/Form/createForm.tsx (hunk 3 above)
Comment (louischan-oursky) [reply src/Form/createForm.tsx]: Your suggestion would be enough. The name should be `scrollToTextInputOffset` though.

Comment (louischan-oursky) [reply src/Form/createForm.tsx]: We should have a default value for this new prop.

**Comment by cychiuae** [comment]
@louischan-oursky updated. THanks
Update start from ba21dce

**Comment by louischan-oursky** [comment]
Great job! Thanks!

**Pull request merged**